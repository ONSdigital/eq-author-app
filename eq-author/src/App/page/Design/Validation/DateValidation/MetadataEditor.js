import React from "react";
import PropTypes from "prop-types";
import { get } from "lodash";

import MetadataContentPicker from "../MetadataContentPicker.js";

const MetadataEditor = ({ onChangeUpdate, answer, validation, readKey }) => (
  <MetadataContentPicker
    answerId={answer.id}
    onSubmit={onChangeUpdate}
    selectedContentDisplayName={get(validation.previousAnswer, "displayName")}
    selectedId={get(validation.previousAnswer, "id")}
    path={`answer.validation.${readKey}.availableMetadata`}
  />
);

MetadataEditor.propTypes = {
  validation: PropTypes.shape({
    id: PropTypes.string.isRequired,
    previousAnswer: PropTypes.shape({
      displayName: PropTypes.string.isRequired,
    }),
  }).isRequired,
  answer: PropTypes.shape({
    id: PropTypes.string.isRequired,
  }).isRequired,
  readKey: PropTypes.string.isRequired,
  onChangeUpdate: PropTypes.func.isRequired,
};

export default MetadataEditor;
