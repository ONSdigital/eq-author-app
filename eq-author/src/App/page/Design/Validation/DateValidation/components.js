import React from "react";
import {Input, Select} from "components/Forms";
import Path from "../path.svg?inline";
import ValidationTitle from "../ValidationTitle";
import ValidationError from "components/ValidationError";
import styled from "styled-components";

const DateInput = styled(Input)`
  width: 12em;
  height: 2.5em;
`;

const ConnectedPath = styled(Path)`
  height: 3.6em;
`;

const RelativePositionSelect = styled(Select)`
  width: 6em;
`;

const RelativePositionText = styled(ValidationTitle)`
  margin-top: 0.5em;
  margin-bottom: 0.5em;
`;

const StartDateText = styled.div`
  margin: 0;
  padding-top: 0.5em;
  height: 2.5em;
`;

const StyledError = styled(ValidationError)`
  justify-content: start;
  width: 60%;
`;

const Now = () =>
  <StartDateText>The date the respondent begins the survey</StartDateText>;

export { StyledError, RelativePositionText, RelativePositionSelect, ConnectedPath, DateInput, Now};
