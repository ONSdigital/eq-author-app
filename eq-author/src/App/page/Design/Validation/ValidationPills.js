import React from "react";
import PropTypes from "prop-types";
import styled from "styled-components";
import { compact, isFunction } from "lodash";

import PillTabs from "components/PillTabs";

import * as entityTypes from "constants/validation-entity-types";

export const Pills = styled(PillTabs)`
  width: 5em;
`;

const PillTabContent = styled.div`
  margin-top: 3em;
  margin-bottom: 0;
  width: 100%;
`;

export const ValidationPills = ({
  entityType,
  onEntityTypeChange,
  PreviousAnswer,
  Metadata,
  Now,
  Custom,
  ...rest
}) => {
  const handleOnChange = (value) =>
    onEntityTypeChange({ name: "entityType", value });

  return (
    <Pills
      value={entityType}
      onChange={handleOnChange}
      options={compact([
        isFunction(Now)
          ? {
              id: "Now",
              title: "Start date",
              render: () => (
                <PillTabContent>
                  <Now {...rest} />
                </PillTabContent>
              ),
            }
          : null,
        isFunction(PreviousAnswer)
          ? {
              id: "PreviousAnswer",
              title: "Previous answer",
              render: () => (
                <PillTabContent>
                  <PreviousAnswer {...rest} />
                </PillTabContent>
              ),
            }
          : null,
        isFunction(Metadata)
          ? {
              id: "Metadata",
              title: "Metadata",
              render: () => (
                <PillTabContent>
                  <Metadata {...rest} />
                </PillTabContent>
              ),
            }
          : null,
        isFunction(Custom)
          ? {
              id: "Custom",
              title: "Custom",
              render: () => (
                <PillTabContent>
                  <Custom {...rest} />
                </PillTabContent>
              ),
            }
          : null,
      ])}
    />
  );
};

ValidationPills.propTypes = {
  entityType: PropTypes.oneOf(Object.values(entityTypes)).isRequired,
  onEntityTypeChange: PropTypes.func.isRequired,
  PreviousAnswer: PropTypes.func,
  Metadata: PropTypes.func,
  Custom: PropTypes.func,
  Now: PropTypes.func,
};
