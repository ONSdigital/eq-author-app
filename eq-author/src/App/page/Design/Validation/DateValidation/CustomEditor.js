import React from "react";
import PropTypes from "prop-types";

import { DateInput } from "./components.js";

const CustomEditor = ({validation, onChange, onUpdate}) => (
  <DateInput
    name="customDate"
    type="date"
    value={validation.customDate}
    onChange={onChange}
    onBlur={onUpdate}
    max="9999-12-30"
    min="1000-01-01"
  />
);

CustomEditor.propTypes = {
  validation: PropTypes.shape({
    customDate: PropTypes.string
  }).isRequired,
  onChange: PropTypes.func.isRequired,
  onUpdate: PropTypes.func.isRequired
};

export default CustomEditor;
