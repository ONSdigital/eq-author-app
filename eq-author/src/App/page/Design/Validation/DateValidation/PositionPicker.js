import React from "react";
import PropTypes from "prop-types";
import { RelativePositionSelect } from "./components";

const RELATIVE_POSITIONS = ["before", "after"];

const PositionPicker = ({value, onChange, onUpdate}) =>
  <RelativePositionSelect
    name="relativePosition"
    value={value}
    onChange={onChange}
    onBlur={onUpdate}
    data-test="relative-position-select"
  >
    {RELATIVE_POSITIONS.map(position => (
      <option key={position} value={position}>
        {position}
      </option>
    ))}
  </RelativePositionSelect>;

PositionPicker.propTypes = {
    value: PropTypes.string.isRequired,
    onChange: PropTypes.func.isRequired,
    onUpdate: PropTypes.func.isRequired
};

export default PositionPicker;
