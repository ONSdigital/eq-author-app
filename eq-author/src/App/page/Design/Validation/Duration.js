import React from "react";
import PropTypes from "prop-types";
import styled from "styled-components";
import { colors } from "constants/theme";

import { Number, Select } from "components/Forms";
import { Grid, Column } from "components/Grid";

import { DAYS, MONTHS, YEARS } from "constants/durations";

export const DurationNumber = styled(Number)`
  width: 100%;

  ${({ hasError }) =>
    hasError &&
    `
    border-color: ${colors.red};
    outline-color: ${colors.red};
    box-shadow: 0 0 0 2px ${colors.red};
    border-radius: 4px;
    margin-bottom: 0;
  `}
`;

const UNITS = [DAYS, MONTHS, YEARS];

const Duration = ({
  name,
  units,
  duration: { value, unit },
  onChange,
  onUpdate,
  hasError,
}) => (
    <Grid>
      <Column cols={2}>
        <DurationNumber
          id={`${name}-value`}
          name={`${name}.value`}
          value={value}
          onChange={onChange}
          onBlur={onUpdate}
          max={99999}
          min={0}
          hasError={hasError}
        />
      </Column>
      <Column cols={4}>
        <Select
          id={`${name}-unit`}
          name={`${name}.unit`}
          value={unit}
          onChange={onChange}
          onBlur={onUpdate}
        >
          {units.map(unit => (
            <option key={unit} value={unit}>
              {unit.toLowerCase()}
            </option>
          ))}
        </Select>
      </Column>
    </Grid>
  );

Duration.propTypes = {
  name: PropTypes.string.isRequired,
  duration: PropTypes.shape({
    value: PropTypes.number,
    unit: PropTypes.oneOf(UNITS).isRequired,
  }).isRequired,
  units: PropTypes.arrayOf(PropTypes.string).isRequired,
  onChange: PropTypes.func.isRequired,
  onUpdate: PropTypes.func.isRequired,
  hasError: PropTypes.object, // eslint-disable-line
};

export default Duration;
