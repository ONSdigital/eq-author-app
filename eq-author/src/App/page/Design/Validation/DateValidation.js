import React from "react";
import PropTypes from "prop-types";
import styled from "styled-components";
import { flowRight, get } from "lodash";

import { Input, Select } from "components/Forms";
import { Grid, Column } from "components/Grid";

import PreviousAnswerContentPicker from "./PreviousAnswerContentPicker";
import MetadataContentPicker from "./MetadataContentPicker";
import { ValidationPills } from "./ValidationPills";
import Path from "./path.svg?inline";
import PathEnd from "./path-end.svg?inline";
import EmphasisedText from "./EmphasisedText";
import AlignedColumn from "./AlignedColumn";
import Duration from "./Duration";

import withChangeUpdate from "enhancers/withChangeUpdate";

import * as entityTypes from "constants/validation-entity-types";
import { DATE, DATE_RANGE } from "constants/answer-types";
import { DAYS, MONTHS, YEARS } from "constants/durations";

const UNITS = [DAYS, MONTHS, YEARS];
const RELATIVE_POSITIONS = ["Before", "After"];

const DateInput = styled(Input)`
  width: 12em;
  height: 2.5em;
`;

const ConnectedPath = styled(Path)`
  height: 3.6em;
`;

const RelativePositionSelect = styled(Select)`
  width: 6em;
`;

const RelativePositionText = styled(EmphasisedText)`
  margin-top: 0.5em;
  margin-bottom: 0.5em;
`;

const StartDateText = styled.div`
  margin: 0;
  padding-top: 0.5em;
  height: 2.5em;
`;

const START_COL_SIZE = 3;
const END_COL_SIZE = 12 - START_COL_SIZE;

const getUnits = ({ format, type }) => {
  if (type === DATE_RANGE) {
    return UNITS;
  }

  if (format === "dd/mm/yyyy") {
    return UNITS;
  }

  if (format === "mm/yyyy") {
    return UNITS.slice(1);
  }

  return UNITS.slice(2);
};

export class UnwrappedDateValidation extends React.Component {
  PreviousAnswer = () => (
    <PreviousAnswerContentPicker
      answerId={this.props.answer.id}
      onSubmit={this.props.onChangeUpdate}
      selectedContentDisplayName={get(
        this.props.validation.previousAnswer,
        "displayName"
      )}
      selectedId={get(this.props.validation.previousAnswer, "id")}
      path={`answer.validation.${this.props.readKey}.availablePreviousAnswers`}
    />
  );

  Metadata = () => (
    <MetadataContentPicker
      answerId={this.props.answer.id}
      onSubmit={this.props.onChangeUpdate}
      selectedContentDisplayName={get(
        this.props.validation.metadata,
        "displayName"
      )}
      path={`answer.validation.${this.props.readKey}.availableMetadata`}
    />
  );

  Now = () => (
    <StartDateText>The date the respondent begins the survey</StartDateText>
  );

  Custom = () => (
    <DateInput
      name="customDate"
      type="date"
      value={this.props.validation.customDate}
      onChange={this.props.onChange}
      onBlur={this.props.onUpdate}
      max="9999-12-30"
      min="1000-01-01"
    />
  );

  render() {
    const {
      validation: { offset, relativePosition, entityType },
      answer: {
        properties: { format },
        type,
      },
      displayName,
      onChange,
      onUpdate,
      onChangeUpdate,
    } = this.props;

    const availableUnits = getUnits({ format, type });

    const validationPills = {
      Metadata: this.Metadata,
      Custom: this.Custom,
    };

    if (type === DATE) {
      validationPills.PreviousAnswer = this.PreviousAnswer;
      validationPills.Now = this.Now;
    }

    return (
      <div>
        <Grid>
          <AlignedColumn cols={START_COL_SIZE}>
            <EmphasisedText>{displayName} is</EmphasisedText>
          </AlignedColumn>
          <Column cols={END_COL_SIZE}>
            <Duration
              name="offset"
              duration={offset}
              units={availableUnits}
              onChange={onChange}
              onUpdate={onUpdate}
            />
          </Column>
        </Grid>
        <Grid>
          <Column cols={START_COL_SIZE}>
            <ConnectedPath />
          </Column>
        </Grid>
        <Grid>
          <AlignedColumn cols={START_COL_SIZE}>
            {type === DATE && (
              <RelativePositionSelect
                name="relativePosition"
                value={relativePosition}
                onChange={onChange}
                onBlur={onUpdate}
                data-test="relative-position-select"
              >
                {RELATIVE_POSITIONS.map(position => (
                  <option key={position} value={position}>
                    {position}
                  </option>
                ))}
              </RelativePositionSelect>
            )}
            {type === DATE_RANGE && (
              <RelativePositionText>{relativePosition}</RelativePositionText>
            )}
            <PathEnd />
          </AlignedColumn>
          <Column cols={9}>
            <ValidationPills
              entityType={entityType}
              onEntityTypeChange={onChangeUpdate}
              {...validationPills}
            />
          </Column>
        </Grid>
      </div>
    );
  }
}

UnwrappedDateValidation.propTypes = {
  validation: PropTypes.shape({
    id: PropTypes.string.isRequired,
    enabled: PropTypes.bool.isRequired,
    customDate: PropTypes.string,
    previousAnswer: PropTypes.shape({
      displayName: PropTypes.string.isRequired,
    }),
    metadata: PropTypes.shape({
      displayName: PropTypes.string.isRequired,
    }),
    offset: PropTypes.shape({
      unit: PropTypes.string.isRequired,
      value: PropTypes.number.isRequired,
    }).isRequired,
    relativePosition: PropTypes.string.isRequired,
    entityType: PropTypes.oneOf(Object.values(entityTypes)).isRequired,
  }).isRequired,
  answer: PropTypes.shape({
    id: PropTypes.string.required,
    properties: PropTypes.shape({
      format: PropTypes.string,
    }).isRequired,
  }).isRequired,
  onToggleValidationRule: PropTypes.func.isRequired,
  onChangeUpdate: PropTypes.func.isRequired,
  onChange: PropTypes.func.isRequired,
  onUpdate: PropTypes.func.isRequired,
  displayName: PropTypes.string.isRequired,
  readKey: PropTypes.string.isRequired,
  testId: PropTypes.string.isRequired,
};

export default flowRight(withChangeUpdate)(UnwrappedDateValidation);
