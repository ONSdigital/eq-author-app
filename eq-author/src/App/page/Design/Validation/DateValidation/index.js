import React from "react";
import PropTypes from "prop-types";
import { flowRight, find } from "lodash";

import { Grid, Column } from "components/Grid";
import { ValidationPills } from "../ValidationPills";
import ValidationTitle from "../ValidationTitle";
import AlignedColumn from "../AlignedColumn";
import Duration from "../Duration";
import { ERR_NO_VALUE } from "constants/validationMessages";
import withChangeUpdate from "enhancers/withChangeUpdate";
import * as entityTypes from "constants/validation-entity-types";
import { DATE } from "constants/answer-types";
import { DAYS, MONTHS, YEARS } from "constants/durations";
import PathEnd from "../path-end.svg?inline";

import {
  StyledError,
  RelativePositionText,
  ConnectedPath,
  Now,
} from "./components";
import PreviousAnswerEditor from "../PreviousAnswerEditor";
import MetadataEditor from "./MetadataEditor";
import CustomEditor from "./CustomEditor";
import PositionPicker from "./PositionPicker.js";

const START_COL_SIZE = 3;
const END_COL_SIZE = 12 - START_COL_SIZE;

const getUnits = format => {
  switch (format) {
    case "dd/mm/yyyy":
      return [DAYS, MONTHS, YEARS];
    case "mm/yyyy":
      return [MONTHS, YEARS];
    default:
      return [YEARS];
  }
};

const UnwrappedDateValidation = ({
  validation,
  answer,
  displayName,
  onChange,
  onUpdate,
  onChangeUpdate,
  readKey,
}) => {
  const availableUnits = getUnits(
    answer.properties.format ? answer.properties.format : "dd/mm/yyyy"
  );

  const hasDurationError = find(validation.validationErrorInfo.errors, error =>
    error.errorCode.includes("ERR_OFFSET_NO_VALUE")
  );

  return (
    <div>
      <Grid>
        <AlignedColumn cols={START_COL_SIZE}>
          <ValidationTitle>{displayName} is</ValidationTitle>
        </AlignedColumn>
        <Column cols={END_COL_SIZE}>
          <Duration
            name="offset"
            duration={validation.offset}
            value={validation.offset.unit}
            units={availableUnits}
            onChange={onChange}
            onUpdate={onUpdate}
            hasError={hasDurationError}
          />
        </Column>
      </Grid>
      <Grid>
        <Column cols={START_COL_SIZE}>
          <ConnectedPath />
        </Column>
        <Column cols={END_COL_SIZE}>
          {hasDurationError && <StyledError>{ERR_NO_VALUE}</StyledError>}
        </Column>
      </Grid>
      <Grid>
        <AlignedColumn cols={START_COL_SIZE}>
          {answer.type === DATE ? (
            <PositionPicker
              value={validation.relativePosition}
              onChange={onChange}
              onUpdate={onUpdate}
            />
          ) : (
            <RelativePositionText>
              {validation.relativePosition.toLowerCase()}
            </RelativePositionText>
          )}
          <PathEnd />
        </AlignedColumn>
        <Column cols={9}>
          <ValidationPills
            entityType={validation.entityType}
            onEntityTypeChange={onChangeUpdate}
            Metadata={MetadataEditor}
            answer={answer}
            validation={validation}
            readKey={readKey}
            onChange={onChange}
            onUpdate={onUpdate}
            onChangeUpdate={onChangeUpdate}
            Custom={CustomEditor}
            {...(answer.type === DATE
              ? {
                  PreviousAnswer: PreviousAnswerEditor,
                  Now: Now,
                }
              : {})}
          />
        </Column>
      </Grid>
    </div>
  );
};

UnwrappedDateValidation.propTypes = {
  validation: PropTypes.shape({
    id: PropTypes.string.isRequired,
    enabled: PropTypes.bool.isRequired,
    customDate: PropTypes.string,
    previousAnswer: PropTypes.shape({
      displayName: PropTypes.string.isRequired,
    }),
    metadata: PropTypes.shape({
      displayName: PropTypes.string.isRequired,
    }),
    offset: PropTypes.shape({
      unit: PropTypes.string.isRequired,
      value: PropTypes.number,
    }).isRequired,
    relativePosition: PropTypes.string.isRequired,
    entityType: PropTypes.oneOf(Object.values(entityTypes)).isRequired,
  }).isRequired,
  answer: PropTypes.shape({
    id: PropTypes.string.required,
    properties: PropTypes.shape({
      format: PropTypes.string,
    }).isRequired,
  }).isRequired,
  onToggleValidationRule: PropTypes.func.isRequired,
  onChangeUpdate: PropTypes.func.isRequired,
  onChange: PropTypes.func.isRequired,
  onUpdate: PropTypes.func.isRequired,
  displayName: PropTypes.string.isRequired,
  readKey: PropTypes.string.isRequired,
  testId: PropTypes.string.isRequired,
};

export default flowRight(withChangeUpdate)(UnwrappedDateValidation);
export { UnwrappedDateValidation };
