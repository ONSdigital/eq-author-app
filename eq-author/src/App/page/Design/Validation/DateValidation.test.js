import React from "react";
import { shallow } from "enzyme";
import { ValidationPills } from "./ValidationPills";
import { UnwrappedDateValidation } from "./DateValidation";
import Duration from "./Duration";
import { DATE, DATE_RANGE } from "constants/answer-types";
import { CUSTOM } from "constants/validation-entity-types";
import { byTestAttr } from "tests/utils/selectors";

const createWrapper = (props, render = shallow) =>
  render(<UnwrappedDateValidation {...props} />);

describe("Date Validation", () => {
  let props, wrapper;
  let onCustomNumberValueChange = jest.fn();
  let onChangeUpdate = jest.fn();
  let onChange = jest.fn();
  let onUpdate = jest.fn();
  let onToggleValidationRule = jest.fn();

  beforeEach(() => {
    props = {
      validation: {
        id: "123",
        enabled: true,
        customDate: "2018-01-01",
        offset: {
          value: 5,
          unit: "Months",
        },
        relativePosition: "Before",
        entityType: CUSTOM,
        validationErrorInfo: {
          errors: [
            {
              errorCode: "ERR_NO_VALUE",
              field: "custom",
              id: "0efd3ed1-8e0d-4b0c-9e39-59010751dbdf",
              type: "validation",
            }
          ]
        },
      },
      answer: {
        id: "1",
        type: DATE,
        properties: {
          format: "YYYY",
        },
      },
      onCustomNumberValueChange: onCustomNumberValueChange,
      onChangeUpdate: onChangeUpdate,
      onChange: onChange,
      onUpdate: onUpdate,
      onToggleValidationRule: onToggleValidationRule,
      displayName: "foobar",
      readKey: "read",
      testId: "test-id",
    };

    wrapper = createWrapper(props);
  });

  it("should render", () => {
    expect(wrapper).toMatchSnapshot();
  });

  it("should render previous answer, metadata, custom and now for date type answer", () => {
    const answer = {
      ...props.answer,
      type: DATE,
    };
    const wrapper = createWrapper({ ...props, answer });
    expect(wrapper.find(ValidationPills)).toMatchSnapshot();
  });

  it("should render metadata and custom for date range type answer", () => {
    const answer = {
      ...props.answer,
      type: DATE_RANGE,
    };
    const wrapper = createWrapper({ ...props, answer });
    expect(wrapper.find(ValidationPills)).toMatchSnapshot();
  });

  it("should trigger update answer validation when the duration value changes", () => {
    const duration = wrapper.find(Duration);
    duration.simulate("change", "event");
    expect(props.onChange).toHaveBeenCalledWith("event");
    duration.simulate("update", "event");
    expect(props.onUpdate).toHaveBeenCalledWith("event");
  });

  it("should trigger update answer validation when the relative position changes", () => {
    const relativePositionField = wrapper.find(
      byTestAttr("relative-position-select")
    );
    relativePositionField.simulate("change", "event");
    expect(onChange).toHaveBeenCalledWith("event");
    relativePositionField.simulate("blur", "event");
    expect(onUpdate).toHaveBeenCalledWith("event");
  });

  it("should trigger update answer validation when the custom value changes", () => {
    const Custom = wrapper.find(ValidationPills).prop("Custom");
    let customDateField = shallow(<Custom />);
    customDateField.simulate("change", "event");
    expect(onChange).toHaveBeenCalledWith("event");
    customDateField.simulate("blur", "event");
    expect(onUpdate).toHaveBeenCalledWith("event");
  });

  it("should correctly handle previous answer", () => {
    const previousAnswer = {
      id: 1,
    };
    const PreviousAnswer = wrapper.find(ValidationPills).prop("PreviousAnswer");
    shallow(<PreviousAnswer />).simulate("submit", {
      name: "previousAnswer",
      value: previousAnswer,
    });

    expect(onChangeUpdate).toHaveBeenCalledWith({
      name: "previousAnswer",
      value: { id: 1 },
    });
  });

  it("should correctly handle metadata change", () => {
    const metadata = {
      id: 1,
    };

    const Metadata = wrapper.find(ValidationPills).prop("Metadata");
    shallow(<Metadata />).simulate("submit", {
      name: "metadata",
      value: metadata,
    });
    expect(props.onChangeUpdate).toHaveBeenCalledWith({
      name: "metadata",
      value: metadata,
    });
  });

  it("should correctly render 'now' entity type", () => {
    const Now = wrapper.find(ValidationPills).prop("Now");
    expect(shallow(<Now />)).toMatchSnapshot();
  });

  it("should filter the available units for the mm/yyyy", () => {
    const answer = {
      ...props.answer,
      properties: {
        format: "mm/yyyy",
      },
    };
    const wrapper = createWrapper({ ...props, answer });
    const duration = wrapper.find(Duration);
    expect(duration.prop("units")).toMatchSnapshot();
  });

  it("should filter the available units for the dd/mm/yyyy", () => {
    const answer = {
      ...props.answer,
      properties: {
        format: "dd/mm/yyyy",
      },
    };
    const wrapper = createWrapper({ ...props, answer });
    const duration = wrapper.find(Duration);
    expect(duration.prop("units")).toMatchSnapshot();
  });

  it("should filter the available units for the yyyy", () => {
    const answer = {
      ...props.answer,
      properties: {
        format: "yyyy",
      },
    };
    const wrapper = createWrapper({ ...props, answer });
    const duration = wrapper.find(Duration);
    expect(duration.prop("units")).toMatchSnapshot();
  });

  it("should render all units for date range answers", () => {
    const answer = {
      ...props.answer,
      type: DATE_RANGE,
    };
    const wrapper = createWrapper({ ...props, answer });
    const duration = wrapper.find(Duration);
    expect(duration.prop("units")).toMatchSnapshot();
  });
});
