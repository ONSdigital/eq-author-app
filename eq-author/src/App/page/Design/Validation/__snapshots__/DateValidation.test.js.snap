// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Date Validation should correctly render 'now' entity type 1`] = `
<DateValidation__StartDateText>
  The date the respondent begins the survey
</DateValidation__StartDateText>
`;

exports[`Date Validation should filter the available units for the dd/mm/yyyy 1`] = `
Array [
  "Days",
  "Months",
  "Years",
]
`;

exports[`Date Validation should filter the available units for the mm/yyyy 1`] = `
Array [
  "Months",
  "Years",
]
`;

exports[`Date Validation should filter the available units for the yyyy 1`] = `
Array [
  "Years",
]
`;

exports[`Date Validation should render 1`] = `
<div>
  <Grid
    align="top"
    fillHeight={true}
  >
    <AlignedColumn
      cols={3}
    >
      <EmphasisedText>
        foobar
         is
      </EmphasisedText>
    </AlignedColumn>
    <Column
      cols={9}
    >
      <Duration
        duration={
          Object {
            "unit": "Months",
            "value": 5,
          }
        }
        hasError={
          Object {
            "errorCode": "ERR_NO_VALUE",
            "field": "custom",
            "id": "0efd3ed1-8e0d-4b0c-9e39-59010751dbdf",
            "type": "validation",
          }
        }
        name="offset"
        onChange={[MockFunction]}
        onUpdate={[MockFunction]}
        units={
          Array [
            "Years",
          ]
        }
        value="Months"
      />
    </Column>
  </Grid>
  <Grid
    align="top"
    fillHeight={true}
  >
    <Column
      cols={3}
    >
      <DateValidation__ConnectedPath />
    </Column>
    <Column
      cols={9}
    >
      <DateValidation__StyledError>
        Value is required
      </DateValidation__StyledError>
    </Column>
  </Grid>
  <Grid
    align="top"
    fillHeight={true}
  >
    <AlignedColumn
      cols={3}
    >
      <DateValidation__RelativePositionSelect
        data-test="relative-position-select"
        name="relativePosition"
        onBlur={[MockFunction]}
        onChange={[MockFunction]}
        value="Before"
      >
        <option
          key="before"
          value="before"
        >
          before
        </option>
        <option
          key="after"
          value="after"
        >
          after
        </option>
      </DateValidation__RelativePositionSelect>
      <Component />
    </AlignedColumn>
    <Column
      cols={9}
    >
      <ValidationPills
        Custom={[Function]}
        Metadata={[Function]}
        Now={[Function]}
        PreviousAnswer={[Function]}
        entityType="Custom"
        onEntityTypeChange={[MockFunction]}
      />
    </Column>
  </Grid>
</div>
`;

exports[`Date Validation should render all units for date range answers 1`] = `
Array [
  "Days",
  "Months",
  "Years",
]
`;

exports[`Date Validation should render metadata and custom for date range type answer 1`] = `
<ValidationPills
  Custom={[Function]}
  Metadata={[Function]}
  entityType="Custom"
  onEntityTypeChange={[MockFunction]}
/>
`;

exports[`Date Validation should render previous answer, metadata, custom and now for date type answer 1`] = `
<ValidationPills
  Custom={[Function]}
  Metadata={[Function]}
  Now={[Function]}
  PreviousAnswer={[Function]}
  entityType="Custom"
  onEntityTypeChange={[MockFunction]}
/>
`;
